rest_framework_json_api/renderers.py

```diff
@@ -610,20 +614,11 @@ def render(self, data, accepted_media_type=None, renderer_context=None):
         else:
             render_data['data'] = json_api_data
 
-        if len(json_api_included) > 0:
-            # Iterate through compound documents to remove duplicates
-            seen = set()
-            unique_compound_documents = list()
-            for included_dict in json_api_included:
-                type_tuple = tuple((included_dict['type'], included_dict['id']))
-                if type_tuple not in seen:
-                    seen.add(type_tuple)
-                    unique_compound_documents.append(included_dict)
-
-            # Sort the items by type then by id
-            render_data['included'] = sorted(
-                unique_compound_documents, key=lambda item: (item['type'], item['id'])
-            )
+        if included_cache:
+            render_data['included'] = list()
+            for included_type in sorted(included_cache.keys()):
```

Comment:
Calling `.keys()` on this line and the next is not necessary. `sorted` applies to the keys of a dictionary by default.

Reply:
Alternatively, this could be compacted with a nested list comprehension:
```python
render_data['included'] = [
    instance for instance_type, type_values in sorted(instance_cache.items())
             for id, instance in sorted(type_values.items())]
```
This works because sorted on `items` works on the keys. This version eliminates a bunch of append operations. There's probably a small perf gain here, but I bet it's tiny. Here's an example to illustrate:
```python
>>> instance_cache = {'c': {'b': 6, 'a': 5}, 'a': {'b': 2, 'a': 1}, 'b': {'b': 4, 'a': 3}}
>>> [instance for instance_type, type_values in sorted(instance_cache.items()) for id, instance in sorted(type_values.items())]
[1, 2, 3, 4, 5, 6]
```

Reply:
If you don't mind I prefer explicitness in code. Most people know that iterating a dictionary in Python yields keys, but using `keys()` makes it obvious for everyone – seniors and juniors from whatever programming language background.

I feel similarly about that nested list comprehension. I didn't know that a thing like that existed in Python, but I have to say it is very difficult to read. Seems arcane.

Reply:
I can appreciate your point of view here. Thanks for at least considering it.